docs/Specs/StoringInfoInOps.md

```diff
@@ -0,0 +1,631 @@
+        Operation *op = $_op.getOperation();
+        return op->getAttrOfType<StringAttr>("myAttr") )
+      }]>
+  ]
+}
+```
+This interface defines:
+- `setMyAttr`: assigns a string attribute called `myAttr` using `setAttr`.
+- `getMyAttr`: retrieves the `myAttr` attribute, returning a null pointer if it is not present.
+
+To enable an operation to use this interface, list it in the operation definition:
+```tablegen
+def MyOp : Op<"my_op", [MyInterface]> {
+  let arguments = ...
+```
+Now, any `MyOp` operation supports `getMyAttr()` and `setMyAttr(...)` methods, usable in C++:
+```
+if (auto myInterface = llvm::dyn_cast<MyInterface>(op)) {
+  myInterface.setMyAttr("foo");
+}
+```
+
+However, you cannot add custom verification directly to an interface, making interface attribute methods less safe than operation attribute methods. If desired, you could call an interface verify function from the verify function of each operation that implements it, but this scales poorly. 
+
+In this way, interfaces offer attribute abstraction, but not attribute verification.
+
+### Good Example
+
+#### InternalDelay for Arithmetic Ops
+
+This is an example of an interface to store the attribute that saves internal delay information in Dynamatic.
+
+https://github.com/EPFL-LAP/dynamatic/blob/f52cb1f922897faf1b66fb087a601064f89e11b4/include/dynamatic/Dialect/Handshake/HandshakeInterfaces.td#L221-L246
+
+The two functions in the interface set and retrieve the attribute. Rather than returning null attributes, if the attribute is not present, a default value of "0.0" is returned.
+
+The interface is added to Handshake_Arith_Op, the base class for all arithmetic operations.
+
+https://github.com/EPFL-LAP/dynamatic/blob/f52cb1f922897faf1b66fb087a601064f89e11b4/include/dynamatic/Dialect/Handshake/HandshakeArithOps.td#L21-L23
+
+The interface can be used to set the value of the attribute (in Buffer Placement) or to retrieve the value of the attribute (in the HandshakeToHw conversion pass):
+
+https://github.com/EPFL-LAP/dynamatic/blob/f52cb1f922897faf1b66fb087a601064f89e11b4/lib/Transforms/BufferPlacement/HandshakePlaceBuffers.cpp#L166-L175
+
+https://github.com/EPFL-LAP/dynamatic/blob/f52cb1f922897faf1b66fb087a601064f89e11b4/lib/Conversion/HandshakeToHW/HandshakeToHW.cpp#L693-L697 
+
+As shown in this example, this interface can be used across different passes and at different stages of the flow.
+
+## Free Functions
+
+If the presence or absence of an attribute is independent of an operation's type, interfaces may not make sense, as the interface would need to be added to every single operation. Instead getters and setters can be implemented as free functions. Free functions are functions which are not part of any class.
+
+Free verification functions could also be written if desired, but like with interfaces, they would need to be manually added to the verify function of each individual operation, which makes verification more-or-less impossible.
```

Comment:
So I guess this is "unlike with interfaces".

In general, I think we should avoid free functions as much as possible. In fact, all operations in the Handshake dialect inherit from the `Handshake_Op` class defined in TableGen. Some _global_ interfaces like `NamedIOInterface`, which provides channel names for RTL generation, are implemented there:

https://github.com/EPFL-LAP/dynamatic/blob/7532a8e33cd3acf4196fece75a4dc20a60b66bb3/include/dynamatic/Dialect/Handshake/HandshakeOps.td#L29-L35

So we can still use interfaces and implement them in the base op class of the dialect.

Maybe going into TableGen classes is a bit too much for this document, making it the complete MLIR Dialect guide😂